### PR TITLE
Cover Infrastructure creation with multiple NAT IP names

### LINKED ISSUE #317: Cover Infrastructure creation with multiple `.networks.cloudNAT.natIPNames` with (env) test
**How to categorize this issue?**

/area testing
/kind enhancement
/priority 3
/platform gcp

**What would you like to be added**:
Previously we had a regression (https://github.com/gardener/gardener-extension-provider-gcp/issues/265) on creating Infrastructure with more than 1 `.networks.cloudNAT.natIPNames`. It was fixed by #266 but no test was added.

**Why is this needed**:
To make sure that the functionality won't regress in the feature.

### PR DESCRIPTION
**Pull request by dimitar-kostadinov**
**How to categorize this PR?**

/area testing
/kind test
/platform gcp

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #317 

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: test/integration/infrastructure/infrastructure_test.go [hunk 1]
```diff
@@ -551,6 +598,24 @@ func verifyCreation(
 	Expect(routerNAT.Subnetworks[0].Name).To(Equal(subnetNodes.SelfLink))
 	Expect(routerNAT.Subnetworks[0].SourceIpRangesToNat).To(Equal([]string{"ALL_IP_RANGES"}))
 
+	if providerConfig.Networks.CloudNAT != nil && len(providerConfig.Networks.CloudNAT.NatIPNames) > 0 {
+		Expect(routerNAT.NatIpAllocateOption).To(Equal("MANUAL_ONLY"))
+
+		// ip addresses
+		var ipAddresses = make(map[string]bool)
+		var address *compute.Address
```
Comment (ialidzhikov) [inline test/integration/infrastructure/infrastructure_test.go:606]: Can't you actually use 

```
			address, err := computeService.Addresses.Get(project, *region, natIPName.Name).Context(ctx).Do()
```

and skip declaring the address var?

File: test/integration/infrastructure/infrastructure_test.go [hunk 2]
```diff
@@ -551,6 +598,24 @@ func verifyCreation(
...
+		for _, natIPName := range providerConfig.Networks.CloudNAT.NatIPNames {
+			address, err = computeService.Addresses.Get(project, *region, natIPName.Name).Context(ctx).Do()
+			Expect(err).NotTo(HaveOccurred())
+			ipAddresses[address.SelfLink] = true
+		}
+		for _, natIP := range routerNAT.NatIps {
+			Expect(ipAddresses).Should(HaveKey(natIP))
+		}
+	} else {
+		Expect(routerNAT.NatIpAllocateOption).To(Equal("AUTO_ONLY"))
```
Comment (ialidzhikov) [inline test/integration/infrastructure/infrastructure_test.go:616]:
```suggestion
		Expect(routerNAT.NatIpAllocateOption).To(Equal("AUTO_ONLY"))
		Expect(routerNAT.NatIps).To(BeEmpty())
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 3]
```diff
@@ -193,24 +193,39 @@ var _ = Describe("Infrastructure tests", func() {
...
 					Name: cloudRouterName,
 				},
-			})
+			}
+			var natIPNames []gcpv1alpha1.NatIPName
+			for _, ipAddressName := range ipAddressNames {
+				natIPNames = append(natIPNames, gcpv1alpha1.NatIPName{Name: ipAddressName})
+			}
+			cloudNAT := &gcpv1alpha1.CloudNAT{
+				MinPortsPerVM: pointer.Int32(2048),
```
Comment (ialidzhikov) [inline test/integration/infrastructure/infrastructure_test.go:225]: Do we have to set it as part of https://github.com/gardener/gardener-extension-provider-gcp/issues/317?
You can either remove this or add assertions that verify later that this field is reflected in the created infrastructure on the cloud provider side.

File: test/integration/infrastructure/infrastructure_test.go [hunk 4]
```diff
@@ -451,21 +469,70 @@ func teardownNetwork(ctx context.Context, logger logr.Logger, project string, co
 
 	routerOp, err := computeService.Routers.Delete(project, *region, routerName).Context(ctx).Do()
 	if err != nil {
-		return err
-	}
-
-	logger.Info("Waiting until router is deleted...", "router", routerName)
-	if err := waitForOperation(ctx, project, computeService, routerOp); err != nil {
-		return err
+		if isNotFoundError(err) {
```
Comment (ialidzhikov) [inline test/integration/infrastructure/infrastructure_test.go:472]: Cosmetic suggestion:

```suggestion
		if gcpclient.IsErrorCode(err, http.StatusNotFound) {
```

You could reuse the already existing func https://github.com/gardener/gardener-extension-provider-gcp/blob/bbaf3eea152e6192d3c1bb7a735a9627e04ca5d3/pkg/gcp/client/errors.go#L21-L39 and drop the `isNotFoundError` func.

File: test/integration/infrastructure/infrastructure_test.go (hunk 1 above)
Comment (ialidzhikov) [reply test/integration/infrastructure/infrastructure_test.go:626]: This suggestion is not applied. Do I miss something that we discussed?

**Review by ialidzhikov — DISMISSED** [review]
Thanks!

/lgtm

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by dkistner — APPROVED** [review]
/lgtm

Thank you @dimitar-kostadinov for catching this up.
@kon-angelo You might want to consider this for the next release already?

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**